### ISSUE TITLE
kubelet has broken ECR auth configuration

### ISSUE THREAD

**Issue by nairb774** [aws/eks-distro#129]
**What happened**:

Setting `AWS_PROFILE` and `AWS_CONFIG_FILE` in an effort to change the role the kubelet uses to authenticate with AWS services didn't impact how ECR authentication was performed.

**What you expected to happen**:

In an effort to minimize the amount of implicit permissions available to pode on EKS, I moved the kubelet's permissions to a separate role, and instructed the kubelet to assume this new role by updating the relevant systemd unit. Most things seemed to work, but ECR auth didn't seem to be working like everything else. Confused with the odd behavior, as I expected https://github.com/kubernetes/kubernetes/blob/v1.17.12/pkg/credentialprovider/aws/aws_credentials.go#L215-L221 to be the relevant code, I took to decompiling the kubelet available in the provided EKS AMI. To my surprise, there was additional code added which doesn't seem to be documented anywhere: https://github.com/aws/eks-distro/blob/91374b70c163875908715a225fb9bdbb90b10d5b/projects/kubernetes/kubernetes/1-18/patches/0006-EKS-PATCH-using-regional-sts-endpoint-for-assume-ecr.patch#L74-L100 While I haven't yet updated to 1.18, the assembly I found in the Kubelet looks fairly similar to that code. The environment variable `ECR_PULL_ROLE_ARN` seems to be completely unknown to the greater internet (no results come up on Google, Bing or DuckDuckGo), and it would be really appreciated to either have this code properly documented, or ideally, just leave the default credential provider alone so all of the environment variable settings work as one might expect.

**How to reproduce it (as minimally and precisely as possible)**:
Run the following on top of the current managed node group AMI (snippeted from our packer config):

```shell
cat > /etc/systemd/system/kubelet.service.d/aws-profile.conf <<"EOF"
[Service]
# The following config file needs to be created before attempting to start the
# kubelet:
EnvironmentFile=/etc/kubernetes/kubelet/aws.env
EOF
# Small kludge: We need to have the /usr/bin/aws-iam-authenticator (used for
# authenticating the kubelet with the control plane) use the default profile due
# to how the kube-system/aws-auth ConfigMap is managed. This makes a poor
# assumption about the format, but what can you do? We capture the leading
# indentation, and use that for forming the env-var block, hoping that the block
# doesn't exist.
! grep 'env:' /var/lib/kubelet/kubeconfig # Verify assumption
sed -i -e 's/^\( *\)command:.*$/&\
\1env:\
\1- name: AWS_PROFILE\
\1  value: default/' /var/lib/kubelet/kubeconfig
# Make sure the replacement happened:
grep -q 'env:' /var/lib/kubelet/kubeconfig
```

Set the contents of `/etc/kubernetes/kubelet/aws.env` to something like the following (easily done with cloud-init):

```
AWS_CONFIG_FILE=/etc/kubernetes/kubelet/aws.conf
AWS_PROFILE=kubelet
```

And set `/etc/kubernetes/kubelet/aws.conf` to the following with the following obvious substitution for the role_arn:
```
[profile kubelet]
credential_source = Ec2InstanceMetadata
role_arn = ${var.roles.kubelet_role_arn}
```

Make sure to setup the correct assume role permissions for the EC2 instance's role, and have the `arn:aws:iam::aws:policy/AmazonEC2ContainerRegistryReadOnly` and `arn:aws:iam::aws:policy/AmazonEKSWorkerNodePolicy` hanging off of the Kubelet role instead.

**Anything else we need to know?**:

The goal of this is to minimize the amount of permissions easily accessible by pods running on the worker node. While this doesn't prevent the pods from assuming the same role as kubelet, it does prevent accidental use of the policies and helps to make sure that each pod has a correctly configured ServiceAccount/IAM Role setup.

**Possible fix:**

Looking at the code, it isn't clear why the setting of `sess.Config.Credentials` is unconditional at https://github.com/aws/eks-distro/blob/91374b70c163875908715a225fb9bdbb90b10d5b/projects/kubernetes/kubernetes/1-18/patches/0006-EKS-PATCH-using-regional-sts-endpoint-for-assume-ecr.patch#L100. It might make more sense to only change that value when the undocumented environment variables are set, thus explicitly choosing this path. Otherwise, the code should ideally behave as similar to what would normally be expected.

**Environment**:
- EKS Distro Release Channel: AWS EKS
- EKS Distro Release: AWS EKS 1.17 eks.4

**Comment by TerryHowe** [comment]
Sounds like you've done a lot of research on this. Do you want to do a PR? I think taking any action 1.17 is a bit late, but it isn't too late to get any changes in 1.19.

https://github.com/aws/eks-distro/blob/91374b70c163875908715a225fb9bdbb90b10d5b/projects/kubernetes/kubernetes/1-19/patches/0006-EKS-PATCH-using-regional-sts-endpoint-for-assume-ecr.patch#L74-L100

I can't say what would make all parties happy with a change there though, but let me know your thoughts.

**Comment by nairb774** [comment]
What isn't clear, and mainly why I didn't propose a PR was there doesn't seem to be any documentation as to why this feature exists, or why it might be needed. Lacking that context, and documentation would it be possible to just remove the patch entirely? With https://github.com/aws/eks-distro/blob/45bbfb32b8b52c6acefcd8b431227aa0d2bfa10e/projects/kubernetes/kubernetes/1-19/patches/0008-EKS-PATCH-Update-aws-sdk-go-to-v1.34.24.patch it looks like the AWS SDK supports the feature directly now: https://github.com/aws/aws-sdk-go/blob/v1.34.24/aws/session/shared_config.go#L285-L292

If the sole purpose was to support regional STS, it looks like this is now addressed. That said, getting any more detail on the reason for the patch would go a long way in knowing the steps forward. The two patches adding the code are the only two files I can find, which makes me suspicious that there aren't any tests covering this feature. If they exist, and anyone knows where they might be, that would be greatly appreciated.

**Comment by TerryHowe** [comment]
Good question, I'm not sure of the answer, but I'll see if anyone else does.

**Comment by nairb774** [comment]
I posted https://github.com/aws/eks-distro/pull/132 to just remove the patch in the off chance that path is possible. If we are able to locate and document a need for the existing code, I can change the PR, or make a new one, to only change the credential loading behavior in the presence of the env vars.

**Comment by SaranBalaji90** [comment]
@nairb774 you were right that this is not required anymore given that aws-sdk-go dependency was updated (https://github.com/kubernetes/kubernetes/pull/87253) from 1.18 onwards and was only added to support pulling images using different role for ECR to support other computes like Fargate and also to support regional STS endpoint which are now addressed. Having said that, we will remove this in the next release. Sorry for the inconvenience this has caused.

**Comment by nairb774** [comment]
Thanks for the background and great to know the history behind the change.

**Comment by TerryHowe** [comment]
Sorry, after a lot of arm twisting I still couldn't get this in v1.19, but it will definitely be in v1.20

**Comment by nairb774** [comment]
It happens. On the other hand, excited to see 1.19 coming.

**Comment by nairb774** [comment]
To follow up, this was fixed in 1.20? I haven't yet been able to test, but curious for when we get there.

**Comment by TerryHowe** [comment]
This was fixed for 1.20
